File: tools/generate_logic_svg.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"strconv"
	"strings"
)

// Gate represents a logic gate or input/output in the circuit
type Gate struct {
	ID    string `json:"id"`
	Type  string `json:"type"`
	Label string `json:"label,omitempty"`
	In1   string `json:"in1,omitempty"`
	In2   string `json:"in2,omitempty"`
	X     int    `json:"x"`
	Y     int    `json:"y"`
}

// Circuit represents the entire circuit definition
type Circuit struct {
	Gates []Gate `json:"gates"`
}

const (
	gateWidth  = 80
	gateHeight = 40
	pinRadius  = 3
	fontSize   = 16
)

func generateSVG(circuit Circuit) (string, error) {
	var svgElements []string
	maxX := 0
	maxY := 0

	// Calculate overall SVG dimensions
	for _, gate := range circuit.Gates {
		if gate.X+gateWidth > maxX {
			maxX = gate.X + gateWidth
		}
		if gate.Y+gateHeight > maxY {
			maxY = gate.Y + gateHeight
		}
	}

	width := maxX + 100 // Add some margin
	height := maxY + 100 // Add some margin

	// Store gate output coordinates
	gateOutputCoords := make(map[string][2]int) // Stores [x, y] for the output pin of each gate

	// First pass: Draw gates and determine output coordinates
	for _, gate := range circuit.Gates {
		x := gate.X
		y := gate.Y

		switch gate.Type {
		case "INPUT":
			svgElements = append(svgElements, fmt.Sprintf(`<text x="%d" y="%d" font-family="monospace" font-size="%d">%s</text>`, x-20, y+gateHeight/2+fontSize/2-5, fontSize, gate.Label))
			outputPinX := x + gateWidth
			outputPinY := y + gateHeight/2
			svgElements = append(svgElements, fmt.Sprintf(`<circle cx="%d" cy="%d" r="%d" fill="black"/>`, outputPinX, outputPinY, pinRadius))
			gateOutputCoords[gate.ID+".out"] = [2]int{outputPinX, outputPinY}
		case "OUTPUT":
			inputPinX := x
			inputPinY := y + gateHeight/2
			svgElements = append(svgElements, fmt.Sprintf(`<circle cx="%d" cy="%d" r="%d" fill="black"/>`, inputPinX, inputPinY, pinRadius))
			svgElements = append(svgElements, fmt.Sprintf(`<text x="%d" y="%d" font-family="monospace" font-size="%d">%s</text>`, x+gateWidth+10, y+gateHeight/2+fontSize/2-5, fontSize, gate.Label))
		default: // Standard logic gates (XOR, AND, OR)
			svgElements = append(svgElements, fmt.Sprintf(`<rect x="%d" y="%d" width="%d" height="%d" fill="lightblue" stroke="black"/>`, x, y, gateWidth, gateHeight))
			svgElements = append(svgElements, fmt.Sprintf(`<text x="%d" y="%d" text-anchor="middle" font-family="monospace" font-size="%d">%s</text>`, x+gateWidth/2, y+gateHeight/2+fontSize/2-5, fontSize, gate.Type))

			// Input pins
			svgElements = append(svgElements, fmt.Sprintf(`<circle cx="%d" cy="%d" r="%d" fill="black"/>`, x, y+gateHeight/4, pinRadius))   // Input 1
			svgElements = append(svgElements, fmt.Sprintf(`<circle cx="%d" cy="%d" r="%d" fill="black"/>`, x, y+3*gateHeight/4, pinRadius)) // Input 2

			// Output pin
			outputPinX := x + gateWidth
			outputPinY := y + gateHeight/2
			svgElements = append(svgElements, fmt.Sprintf(`<circle cx="%d" cy="%d" r="%d" fill="black"/>`, outputPinX, outputPinY, pinRadius))
			gateOutputCoords[gate.ID+".out"] = [2]int{outputPinX, outputPinY}
		}
	}

	// Second pass: Draw connections
	for _, gate := range circuit.Gates {
		x := gate.X
		y := gate.Y

		// Connect input 1
		if gate.In1 != "" && gate.Type != "INPUT" {
			sourceRef := gate.In1
			if coords, ok := gateOutputCoords[sourceRef]; ok {
				sx, sy := coords[0], coords[1]
				dx, dy := x, y+gateHeight/4 // Input 1 pin
				svgElements = append(svgElements, fmt.Sprintf(`<line x1="%d" y1="%d" x2="%d" y2="%d" stroke="black" stroke-width="2"/>`, sx, sy, dx, dy))
			}
		}

		// Connect input 2
		if gate.In2 != "" && gate.Type != "INPUT" {
			sourceRef := gate.In2
			if coords, ok := gateOutputCoords[sourceRef]; ok {
				sx, sy := coords[0], coords[1]
				dx, dy := x, y+3*gateHeight/4 // Input 2 pin
				svgElements = append(svgElements, fmt.Sprintf(`<line x1="%d" y1="%d" x2="%d" y2="%d" stroke="black" stroke-width="2"/>`, sx, sy, dx, dy))
			}
		}
	}

	svgContent := fmt.Sprintf(`<svg width="%d" height="%d" viewBox="0 0 %d %d" xmlns="http://www.w3.org/2000/svg">
    %s
</svg>`, width, height, width, height, strings.Join(svgElements, "\n    "))

	return svgContent, nil
}

func main() {
	if len(os.Args) < 2 {
		fmt.Println("Usage: go run main.go <json_file_path>")
		os.Exit(1)
	}

	jsonFilePath := os.Args[1]
	data, err := os.ReadFile(jsonFilePath)
	if err != nil {
		fmt.Printf("Error reading file: %v\n", err)
		os.Exit(1)
	}

	var circuit Circuit
	err = json.Unmarshal(data, &circuit)
	if err != nil {
		fmt.Printf("Error unmarshaling JSON: %v\n", err)
		os.Exit(1)
	}

	svgOutput, err := generateSVG(circuit)
	if err != nil {
		fmt.Printf("Error generating SVG: %v\n", err)
		os.Exit(1)
	}

	fmt.Println(svgOutput)
}

File: tools/generate_logic_svg.py
import json
import sys

def generate_svg(gate_data):
    svg_elements = []
    gate_width = 80
    gate_height = 40
    pin_radius = 3
    font_size = 16

    # Calculate overall SVG dimensions based on gate positions
    max_x = 0
    max_y = 0
    for gate in gate_data['gates']:
        max_x = max(max_x, gate.get('x', 0) + gate_width)
        max_y = max(max_y, gate.get('y', 0) + gate_height)

    width = max_x + 100 # Add some margin
    height = max_y + 100 # Add some margin

    # Store gate positions and output coordinates
    gate_output_coords = {} # Stores (x, y) for the output pin of each gate

    # First pass: Draw gates and determine output coordinates
    for gate in gate_data['gates']:
        gate_id = gate['id']
        x = gate.get('x', 0)
        y = gate.get('y', 0)
        gate_type = gate['type']
        label = gate.get('label', gate_type)

        if gate_type == 'INPUT':
            svg_elements.append(f'<text x="{x - 20}" y="{y + gate_height / 2 + font_size / 2 - 5}" font-family="monospace" font-size="{font_size}">{label}</text>')
            # Output pin for INPUT gates is on the right
            output_pin_x = x + gate_width
            output_pin_y = y + gate_height / 2
            svg_elements.append(f'<circle cx="{output_pin_x}" cy="{output_pin_y}" r="{pin_radius}" fill="black"/>')
            gate_output_coords[gate_id + '.out'] = (output_pin_x, output_pin_y)
        elif gate_type == 'OUTPUT':
            # Input pin for OUTPUT gates is on the left
            input_pin_x = x
            input_pin_y = y + gate_height / 2
            svg_elements.append(f'<circle cx="{input_pin_x}" cy="{input_pin_y}" r="{pin_radius}" fill="black"/>')
            svg_elements.append(f'<text x="{x + gate_width + 10}" y="{y + gate_height / 2 + font_size / 2 - 5}" font-family="monospace" font-size="{font_size}">{label}</text>')
        else: # Standard logic gates (XOR, AND, OR)
            svg_elements.append(f'<rect x="{x}" y="{y}" width="{gate_width}" height="{gate_height}" fill="lightblue" stroke="black"/>')
            svg_elements.append(f'<text x="{x + gate_width / 2}" y="{y + gate_height / 2 + font_size / 2 - 5}" text-anchor="middle" font-family="monospace" font-size="{font_size}">{gate_type}</text>')

            # Input pins
            svg_elements.append(f'<circle cx="{x}" cy="{y + gate_height / 4}" r="{pin_radius}" fill="black"/>') # Input 1
            svg_elements.append(f'<circle cx="{x}" cy="{y + 3 * gate_height / 4}" r="{pin_radius}" fill="black"/>') # Input 2

            # Output pin
            output_pin_x = x + gate_width
            output_pin_y = y + gate_height / 2
            svg_elements.append(f'<circle cx="{output_pin_x}" cy="{output_pin_y}" r="{pin_radius}" fill="black"/>')
            gate_output_coords[gate_id + '.out'] = (output_pin_x, output_pin_y)

    # Second pass: Draw connections
    for gate in gate_data['gates']:
        gate_id = gate['id']
        gate_x = gate.get('x', 0)
        gate_y = gate.get('y', 0)

        # Connect input 1
        if 'in1' in gate and gate['type'] != 'INPUT':
            source_ref = gate['in1']
            if source_ref in gate_output_coords:
                sx, sy = gate_output_coords[source_ref]
                dx, dy = gate_x, gate_y + gate_height / 4 # Input 1 pin
                svg_elements.append(f'<line x1="{sx}" y1="{sy}" x2="{dx}" y2="{dy}" stroke="black" stroke-width="2"/>')

        # Connect input 2
        if 'in2' in gate and gate['type'] != 'INPUT':
            source_ref = gate['in2']
            if source_ref in gate_output_coords:
                sx, sy = gate_output_coords[source_ref]
                dx, dy = gate_x, gate_y + 3 * gate_height / 4 # Input 2 pin
                svg_elements.append(f'<line x1="{sx}" y1="{sy}" x2="{dx}" y2="{dy}" stroke="black" stroke-width="2"/>')

    svg_content = f'''<svg width="{width}" height="{height}" viewBox="0 0 {width} {height}" xmlns="http://www.w3.org/2000/svg">
    {"".join(svg_elements)}
</svg>'''
    return svg_content

if __name__ == '__main__':
    if len(sys.argv) < 2:
        print("Usage: python generate_logic_svg.py <json_file_path>")
        sys.exit(1)

    json_file_path = sys.argv[1]
    try:
        with open(json_file_path, 'r') as f:
            gate_data = json.load(f)
        svg_output = generate_svg(gate_data)
        print(svg_output)
    except FileNotFoundError:
        print(f"Error: File not found at {json_file_path}")
        sys.exit(1)
    except json.JSONDecodeError:
        print(f"Error: Invalid JSON in {json_file_path}")
        sys.exit(1)
    except Exception as e:
        print(f"An error occurred: {e}")
        sys.exit(1)
